src/topology/metric_space/dilation.lean

```diff
@@ -0,0 +1,274 @@
+lemma to_continuous : continuous (f : α → β) :=
+(lipschitz f).continuous
+
+/-- Similarities multiply the diameter by their ratio in pseudoemetric spaces. -/
+lemma ediam_image (s : set α) :
+  emetric.diam ((f: α → β) '' s) = ratio f * emetric.diam s :=
+begin
+  apply le_antisymm,
+  { exact lipschitz_with.ediam_image_le (lipschitz f) s },
+  by_cases hr : ratio f ≠ 0,
+  { rw [mul_comm, ← ennreal.le_div_iff_mul_le, div_eq_mul_inv, mul_comm, ← ennreal.coe_inv hr],
+    refine antilipschitz_with.le_mul_ediam_image (antilipschitz f hr) s,
+    left, simp [hr],
+    left, simp [hr], },
+  rw not_not at hr,
+  simp [hr],
+end
+
+lemma ediam_range :
+  emetric.diam (range (f : α → β)) = ratio f * emetric.diam (univ : set α) :=
+by { rw ← image_univ, exact ediam_image f univ }
+
+lemma maps_to_emetric_ball  (hr : ratio f ≠ 0) (x : α) (r' : ℝ≥0∞) :
+  maps_to (f : α → β) (emetric.ball x r') (emetric.ball (f x) (ratio f * r')) :=
+begin
+  intros y hy,
+  simp only [emetric.mem_ball, edist_eq] at *,
+  rwa ennreal.mul_lt_mul_left _ _,
+  simp [hr],
+  simp [hr],
+end
+
+lemma maps_to_emetric_closed_ball (x : α) (r' : ℝ≥0∞) :
+  maps_to (f : α → β) (emetric.closed_ball x r') (emetric.closed_ball (f x) (ratio f * r')) :=
+begin
+  by_cases hr : ratio f ≠ 0,
+  { intros y hy,
+    simp only [emetric.mem_closed_ball, edist_eq] at *,
+    rwa ennreal.mul_le_mul_left _ _,
+    simp [hr],
+    simp [hr], },
+  rw not_not at hr,
+  simp [hr, maps_to],
+end
+
+lemma comp_continuous_on_iff
+  {γ} [topological_space γ] {g : γ → α} {s : set γ} (hr : ratio f ≠ 0) :
+  continuous_on ((f : α → β) ∘ g) s ↔ continuous_on g s :=
+(uniform_inducing f hr).inducing.continuous_on_iff.symm
+
+lemma comp_continuous_iff
+  {γ} [topological_space γ] {g : γ → α} (hr : ratio f ≠ 0) :
+  continuous ((f : α → β) ∘ g) ↔ continuous g :=
+(uniform_inducing f hr).inducing.continuous_iff.symm
+
+end pseudo_emetric_dilation --section
+
+section emetric_dilation
+variables [emetric_space α]
+
+/-- A dilation from a metric space is a uniform embedding -/
+theorem uniform_embedding [pseudo_emetric_space β] [dilation_class F α β]
+  (f : F) (hr : ratio f ≠ 0) : uniform_embedding f :=
+(antilipschitz f hr).uniform_embedding (lipschitz f).uniform_continuous
+
+/-- A dilation from a metric space is an embedding -/
+theorem embedding [pseudo_emetric_space β] [dilation_class F α β]
+  (f : F) (hr : ratio f ≠ 0) : embedding (f : α → β) :=
+(uniform_embedding f hr).embedding
+
+/-- A dilation from a complete emetric space is a closed embedding -/
+theorem closed_embedding
+  [complete_space α] [emetric_space β] [dilation_class F α β]
+  (f : F) (hr : ratio f ≠ 0) : closed_embedding (f : α → β) :=
+(antilipschitz f hr).closed_embedding (lipschitz f).uniform_continuous
+
+end emetric_dilation --section
+
+section pseudo_metric_dilation
+variables [pseudo_metric_space α] [pseudo_metric_space β] [dilation_class F α β] (f : F)
+
+/-- An isometry preserves the diameter in pseudometric spaces. -/
```

Comment:
Doc string should be corrected to describe the effects of a dilation not an isometry.